File: src/shared/py/langboard_shared/domain/models/ChatGraphApprovalRequest.py
from typing import Any, Self
from ...core.db import SnowflakeIDField
from ...core.types import SnowflakeID
from .bases import BaseGraphApprovalRequestModel
from .bases.BaseGraphApprovalRequestModel import graph_approval_request_api_field
from .ChatHistory import ChatHistory
from .ChatSession import ChatSession
from .GraphApprovalRequest import GraphApprovalOriginType, GraphApprovalRequest


class ChatGraphApprovalRequest(BaseGraphApprovalRequestModel, table=True):
    chat_session_id: SnowflakeID = SnowflakeIDField(
        foreign_key=ChatSession,
        nullable=False,
        index=True,
        api_field=graph_approval_request_api_field("chat_session_uid"),
    )
    chat_history_id: SnowflakeID = SnowflakeIDField(
        foreign_key=ChatHistory,
        nullable=False,
        index=True,
        api_field=graph_approval_request_api_field("chat_history_uid"),
    )

    @staticmethod
    def get_request_type() -> GraphApprovalOriginType:
        return GraphApprovalOriginType.Chat

    @classmethod
    def create_for_approval(
        cls, approval: GraphApprovalRequest, scope_table: str, scope_id: SnowflakeID | None, **context: Any
    ) -> Self:
        chat_session = context.get("chat_session")
        chat_history = context.get("chat_history")
        if not isinstance(chat_session, ChatSession) or not isinstance(chat_history, ChatHistory):
            raise ValueError("Chat graph approval request requires chat session and chat history.")

        return cls(
            approval_request_id=approval.id,
            scope_table=scope_table,
            scope_id=scope_id,
            chat_session_id=chat_session.id,
            chat_history_id=chat_history.id,
        )

File: src/socket/src/models/ChatGraphApprovalRequest.ts
import { BigIntColumn, TBigIntString } from "@/core/db/BaseModel";
import SnowflakeID from "@/core/db/SnowflakeID";
import { TGraphApprovalResponse } from "@/models/GraphApprovalRequestTypes";
import BaseGraphApprovalRequestModel from "@/models/bases/BaseGraphApprovalRequestModel";
import { Entity } from "typeorm";

@Entity({ name: "chat_graph_approval_request" })
class ChatGraphApprovalRequest extends BaseGraphApprovalRequestModel {
    @BigIntColumn(false)
    public chat_session_id!: TBigIntString;

    @BigIntColumn(false)
    public chat_history_id!: TBigIntString;

    public toGraphApprovalResponse(projectID?: TBigIntString): TGraphApprovalResponse {
        return {
            ...this.getBaseApiResponse(projectID),
            chat_session_uid: new SnowflakeID(this.chat_session_id).toShortCode(),
            chat_history_uid: new SnowflakeID(this.chat_history_id).toShortCode(),
        };
    }
}

export default ChatGraphApprovalRequest;
